File: workspace/2023-05-09/string.javascript/string.js

function range(min, max) {
    return Array.from(Array(max - min + 1).keys()).map(x => x + min)
}

function prefixes(string) {
    return range(0, string.length).map(i => {
        return string.substring(0, i)
    })
}

function is_prefix(string1, string2) {
    return string.startsWith(string2)
}

function proper_prefixes(string) {
    return range(0, string.length - 1).map(i => {
        return string.substring(0, i)
    })
}
function is_proper_prefix(string1, string2) {
    return string1.length != string2.length && string1.startsWith(string2)
}

function suffixes(string) {
    return range(0, string.length).map(i => {
        return string.substring(string.length - i, string.length)
    })
}

function is_suffix(string1, string2) {
    return string1.endsWith(string2)
}

function proper_suffixes(string) {
    return range(0, string.length - 1).map(i => {
        return string.substring(string.length - i, string.length)
    })
}

function is_proper_suffix(string1, string2) {
    return string1.length != string2.length && string1.endsWith(string2)
}

// 先頭 i 文字.
function first_i_chars(string, i) {
    return string.substring(0, i)
}

function intersection(array1, array2) {
    return array1.filter(x => array2.includes(x))
}

// LPS (Longest proper Prefix Suffix) Array.
function lps_array(string) {
    return range(1, string.length).map(i => {
        let substring = string.substring(0, i)
        return Math.max(...(intersection(proper_prefixes(substring), proper_suffixes(substring)).map(_ => _.length)))
    })
}

File: workspace/2023-05-20/string.python/string.py

def prefixes(text):
    return list(map(
        lambda n: text[0:n],
        range(0, len(text) + 1)))


def proper_prefixes(text):
    return list(map(
        lambda n: text[0:n],
        range(0, len(text))))


def suffixes(text):
    return list(map(
        lambda n: text[len(text)-n:],
        range(0, len(text) + 1)))


def proper_suffixes(text):
    return list(map(
        lambda n: text[len(text)-n:],
        range(0, len(text))))


def lps(text):
    proper_prefix_set = set(proper_prefixes(text))
    proper_suffix_set = set(proper_suffixes(text))
    intersection = proper_prefix_set.intersection(proper_suffix_set)

    if not intersection:
        return None
    return sorted(intersection, key=len)[-1]


def lps_array(text):
    return list(map(len, map(lps, prefixes(text)[1:])))

